File: Cpp/1005_Maximize_Sum_Of_Array_After_K_Negations.cpp
class Solution {
public:
    int largestSumAfterKNegations(vector<int>& A, int K) {
        int result = 0;
        int min_value = INT_MAX;
        priority_queue<int, vector<int>, greater<int>> pq;
        for(int i=0; i<A.size(); ++i){
            result += A[i];
            pq.push(A[i]);
            min_value = min(min_value, abs(A[i]));
        }
        
        if(pq.top() >= 0) return result - (K&1 == 1) * 2 * min_value;
        
        for(int i=0; i<K; ++i){
            if(pq.top() > 0) return result - ((K-i)&1 == 1) * 2 * min_value;
            result += -2*pq.top();
            pq.pop();
        }
        
        return result;
    }
};

File: Kotlin/1005_Maximize_Sum_Of_Array_After_K_Negations.kt
class Solution {
    fun largestSumAfterKNegations(nums: IntArray, k: Int): Int {
        var result = 0
        var minValue = Integer.MAX_VALUE
        val pq = PriorityQueue<Int>()
        
        nums.forEach { num ->
            result += num
            minValue = minOf(minValue, Math.abs(num))
            pq.offer(num)
        }
        
        if(pq.peek() >= 0) return result - (k and 1) * 2 * minValue

        for(i in 0 until k) {
            if(pq.isEmpty() || pq.peek() > 0) return result - ((k - i) and 1) * 2 * minValue
            result += -2 * pq.poll()
        }
        
        return result
    }
}